File: EasyRecycle/recycle/views/service.py
from rest_framework import generics, permissions

from recycle.models import Service
from recycle.serializers import ServiceSerializer, CreateServiceSerializer, EditServiceSerializer


# /api/v1/recycle/services
# methods:
#   - get
# returns (success status - 200):
#   [
#     {
#       "id": <int>,
#       "garbage_type_name": <string>,
#       "service_name": <string>,
#       "price_per_kg": <float>,
#       "location_id": <int>
#     },
#     ...
#   ]
class ServicesAPIView(generics.ListAPIView):
	permission_classes = (permissions.AllowAny,)
	queryset = Service.objects.order_by('location')
	serializer_class = ServiceSerializer


# /api/v1/recycle/services/<pk>
# path args:
#   - pk <int>: primary key of Service object
# methods:
#   - get
# returns (success status - 200):
#   {
#     "id": <int>,
#     "garbage_type_name": <string>,
#     "service_name": <string>,
#     "price_per_kg": <float>,
#     "location_id": <int>
#   }
class ServiceDetailsAPIView(generics.RetrieveAPIView):
	permission_classes = (permissions.AllowAny,)
	queryset = Service.objects.all()
	serializer_class = ServiceSerializer


# /api/v1/recycle/services/create
# methods:
#   - post:
#       - garbage_type: string
#       - service_name: string
#       - price_per_kg: float
#       - location: int
# returns (success status - 201):
#   {
#     "id": <int>,
#     "garbage_type": <string>,
#     "service_name": <string>,
#     "price_per_kg": <float>,
#     "location": <int>
#   }
class CreateServiceAPIView(generics.CreateAPIView):
	permission_classes = (permissions.IsAdminUser,)
	queryset = Service.objects.all()
	serializer_class = CreateServiceSerializer


# /api/v1/recycle/services/<pk>/manage
# path args:
#   - pk <int>: primary key of Service object
# methods:
#   - put:
#       - garbage_type: string
#       - service_name: string
#       - price_per_kg: float
#       - location: int
#   - delete
# returns (success status: 204 (on delete), 200 (on update)):
#   {
#     "id": <int>,
#     "garbage_type": <string>,
#     "service_name": <string>,
#     "price_per_kg": <float>,
#     "location": <int>
#   }
class ManageServiceAPIView(generics.UpdateAPIView, generics.DestroyAPIView):
	permission_classes = (permissions.IsAdminUser,)
	queryset = Service.objects.all()
	serializer_class = EditServiceSerializer

File: EasyRecycle.UI/src/services/service.js
import BaseService from "./base";

class ServiceService extends BaseService {

    constructor() {
        super();
        this._URL_SERVICES = this._BASE_URL + '/recycle/services';
    }

    // returns:
    //  [
    //    {
    //      "id": <int>,
    //      "garbage_type_name": <string>,
    //      "service_name": <string>,
    //      "price_per_kg": <float>,
    //      "location_id": <int>
    //    }
    //  ]
    getServices = (handler) => {
        return this.get({url: this._URL_SERVICES}, handler);
    }

    // returns:
    //  {
    //    "id": <int>,
    //    "garbage_type_name": <string>,
    //    "service_name": <string>,
    //    "price_per_kg": <float>,
    //    "location_id": <int>
    //  }
    getService = (id, handler) => {
        return this.get({url: this._URL_SERVICES + "/" + id.toString()}, handler);
    }

    // returns:
    //  {
    //    "id": <int>,
    //    "garbage_type_name": <string>,
    //    "service_name": <string>,
    //    "price_per_kg": <float>,
    //    "location_id": <int>
    //  }
    createService = (garbageType, serviceName, pricePerKg, locationId, handler) => {
        let data = {
            garbage_type: garbageType,
            service_name: serviceName,
            price_per_kg: pricePerKg,
            location: locationId
        };
        this.post({
            url: this._URL_SERVICES + '/create',
            data: data
        }, handler);
    }

    // returns:
    //  {
    //    "id": <int>,
    //    "garbage_type_name": <string>,
    //    "service_name": <string>,
    //    "price_per_kg": <float>,
    //    "location_id": <int>
    //  }
    editService = (id, garbageType, serviceName, pricePerKg, locationId, handler) => {
        let data = {};
        if (garbageType !== null)
        {
            data['garbage_type'] = garbageType;
        }

        if (serviceName !== null)
        {
            data['service_name'] = serviceName;
        }

        if (pricePerKg !== null)
        {
            data['price_per_kg'] = pricePerKg;
        }

        if (locationId !== null)
        {
            data['location'] = locationId;
        }

        this.put({
            url: this._URL_SERVICES + "/" + id.toString() + '/manage',
            data: data
        }, handler);
    }

    cancelService = (id, handler) => {
        this.delete_({
            url: this._URL_SERVICES + "/" + id.toString() + '/manage'
        }, handler);
    }
}

export default new ServiceService();
